Automating field notes using mWater

SamStruthers opened the pull request:
mWater is a survey platform used by CU folks to record field note visits. Data is stored in a datagrid associated with a survey on the mWater portal. These data can be downloaded using an API url (mWater api yml)
This PR has three main files which all live in the folder `mWater_collate`:
- `mWater_pull_collate.rmd`: main file which pulls data from mWater API and breaks it into three main sections: all_notes, sensor_notes, sampling_notes. 

Dataframes: All notes is everything submitted, sensor_notes is only visits where sensor was interacted with, sampling notes is when water sampling occured.  

Actions: save sensor notes to csv in data folder this is how we can maintain a backup of the data from mWater. This code also uses the photo_download and sampling spreadsheet creator functions (described below)
- `photo_downloader.R`: grabs all the download urls from the all notes sheet and will download+rename all the photos uploaded to mWater. It should not download any photos if they have already been downloaded
- `sampling_spreadsheet_creator.R`: formats a days worth of water sampling data to match RMRS formatting. This is mainly a tool that I will use after a field day to speed up sample input and make sure I don't create any typos in the RMRS masterfile. 

Review Requested:
- I will send you the API yml. Please test to make sure the whole pipeline works by running all the chunks (keeping in mind that photo downloader wont do anything since I already downloaded the photos)
- The code is not very concise, especially photo downloader so any suggestions on that would be helpful
- Check to make sure I mimic'ed the existing structure of our field notes in QAQC pipeline (start_dt, round_dt, sensor employed, etc)
- general notes on conciseness of rmd file and any sections that seem confusing

juandlt-csu commented on .gitignore:
I was able to use the .yml file for its intended purpose

juandlt-csu commented on src/mWater_collate/download_pictures.R:
```diff
@@ -0,0 +1,107 @@
+
+
+  # loop thru dataset and download the photo ONLY if it is not yet downloaded and not NA
+  for (i in 1:length(sampling_photos$site)) {
```
suggestion: If this is iterating over the number of rows in the df, I think it would be more clear to use `for (i in 1:nrow(sampling_photos))`.

SamStruthers replied:
Will do, thanks!

juandlt-csu commented on src/mWater_collate/mWater_pull_collate.Rmd:
```diff
@@ -0,0 +1,166 @@
+library(tidyverse)
+
+#grab site metadata
+sampling_sites <- read_csv("data/water_sampling_sites.csv")
```
I had to change the file location here, but I don't know why... Generally I was struggling with file locations. I don't know if this is a problem with this PR or with my local instance. Will have to check this again once it is merged.

kathryn-willi replied:
I also had issues with the file locations... not sure why...

juandlt-csu commented on src/mWater_collate/mWater_pull_collate.Rmd:
I think that this is generally working correctly, but I would appreciate some documentation that explains overall what should be happening.

juandlt-csu commented on src/mWater_collate/mWater_pull_collate.Rmd:
````diff
@@ -0,0 +1,166 @@
+## ALL NOTES
+```{r}
+#read in file
+all_notes <- read_csv(file = "data/mWater_raw_sensor_notes.csv")%>%
````
Which one of these dfs are we hoping to append to the field notes sheet that we already have?

SamStruthers replied:
sensor_notes
apologies that was not clear

juandlt-csu commented on src/mWater_collate/mWater_pull_collate.Rmd:
```diff
@@ -0,0 +1,166 @@
+                                    sensor_change %in% c("Swapped", "Deployed") ~ 1), 
+         # Create basis for calibration report name
+         # this will be used to check for calibration report in data files and then b
+         cal_report_name = case_when(cal_report_collected == TRUE ~ paste0(site, "_", format(start_dt, "%Y%m%d")),
```
question: what is the log stuff doing? Does mWater have a way to download and interact with the calibration report stuff?

SamStruthers replied:
Not yet! I want to build out a way to make sure that techs/us are uploading all our files in a timely manner. This will essentially be the file that R will look for in the data folder to see if the file has been uploaded or not

juandlt-csu commented on src/mWater_collate/sampling_spreadsheet_creator.R:
```diff
@@ -0,0 +1,21 @@
+sampling_spreadsheet_creator <- function(date_oi = "2023-10-16" ){
+
+  #pull in site meta data
```
quibble: I like to add lines between steps/chunks of code to improve readability.

juandlt-csu reviewed (approved):
This seems to be working correctly to me. Although, I thought that the end goal would be to generate a df from the mWater field notes that are pulled such that it can be appended to the field notes that we already have. It wasn't clear to me which generated df would be the one that would be appended to our historical field notes.

SamStruthers commented:
Thanks for the notes. I'll make some changes to the scripts to make it clear that sensor_notes is what we will be pulling into the QAQC script.

SamStruthers commented:
Big Improvement (IMO): 
- Improved API pull (no raw mWater file anymore)
- all_notes_cleaned is the df that is used downstream and only has a few minor corrections from API pull
- readability is improved and sensor notes is the only other dataframe created
- Finished log/cal report check function: This checks the data folder to see if users have uploaded the logs/cal reports that they claim to in the mWater survey
- Moved all water sampling dataframes into the sampling spreadsheet function as they are not needed in the main rmd. Users can now save all the water sampling info as csv or just a day in RMRS formatting
- Fixed the nrow comment in download pictures.R

kathryn-willi reviewed (approved):
This is awesome, Sam! And going to make field days much easier. @juandlt-csu, we will need to come up with a way of appending our old field notes to this newer data at the top of our field notes script in the QAQC workflow. We will also want to confirm all the original column names line up. On that topic, I will go ahead and add in a `visit_type` column to our old data and fill that column in so that we can use that field to filter/remove visits where the sensor wasn't actually handled.

SamStruthers commented:
I added a new file `scratch/data_tidying/timelapse_folder_rename.R`
This is mostly an tool that I will use when I get SD cards from our timelapse cameras. 
It essentially prompts the user for the site name after showing a picture from the folder in the `timelapse_raw` folder. Based on the response and the first and last photo in the folder, it renames the folder to match our syntax and then moves said folder to the `timelapse_photos` folder. 

No review required since I will be the main user of this and it does not affect our larger workflow unless anyone feels strongly about this....
